[MRG+2] Label K-Fold

glouppe opened the pull request:
This supersedes #4444

JeanKossaifi committed:
Added subject independent KFold

Changed SubjectIndependentKFold to DisjointGroupKFold

cosmetic changes  test (fix seed correctly, use assert_equal for
meaningful error messages)

Changed name to DisjointLabelKFold

Added example of use

FIX: whitespace related doctest failure

FIX: Python 2.6 requires the field numbers in print

FIX: change docstring to comment in test function

DOC: moved docstring from function to class

FIX: added call to parent class

FIX: error in calling the parent

DOC: fixed doctest

FIX: doctest

Cosmetic changes (minor refactoring)

Optimised code (use np.bincount)

Cosmetic: use samples instead of weight for clarity

Minor fix: removed shuffle parameter

Cosmetic

Use mergesort instead of quicksort for reproducibility.

Changed variable name 'y' to 'label'.

Added test for degenerate case where n_folds > n_labels.

Documented the requirement n_labels > n_folds.

DOC: improved description + added see also sections.

Fixed dtype of temporary arrays.

Improved test: check that one label is not in both test and training.

Added documentation for DisjoinLabelKFold.

glouppe committed:
ENH: rename to LabelKFold

COSMIT: doc, pep8, etc

Refactor code

glouppe commented:
I am switching this back to WIP, as I am not sure the `label_kfold` function, in addition to `LabelKFold` class is needed. I will refactor that to follow more closely the structure of the other iterators.

JeanKossaifi commented:
About the name, the reason we had decided disjointLabelKFold and not simply LabelKFold is that it is more informative. In that instance the goal was to perform subject/label independent experiment. It doesn't really matter whether the method acts on labels or something else, the main point is that the resulting folds are disjoint. Isn't it clearer to the user if the name indicates that?

glouppe commented:
I removed the standalone function and moved its logic inside `LabelKFold.__init__`. This is ready for review. I kept @JeanKossaifi original code as much as possible.

For the naming, I have no string opinion, but we should aim for something consistent across all CV iterators. For now "label" appears to be the used terminology, but I agree that "grouped" is maybe clearer.

JeanKossaifi commented:
Seems good to me.

arjoly commented on doc/modules/cross_validation.rst:
```diff
@@ -261,6 +261,33 @@ two slightly unbalanced classes::
+
+  >>> labels = [1, 1, 1, 2, 2, 2, 3, 3, 3, 3]
+
+  >>> lkf = LabelKFold(labels, 3)
```
I would use `lkf = LabelKFold(labels, n_folds=3)`.

arjoly commented on sklearn/cross_validation.py:
```diff
@@ -339,6 +342,121 @@ def __len__(self):
+        super(LabelKFold, self).__init__(len(labels), n_folds, shuffle,
+                                         random_state)
+
+        self.n = len(labels)
```
one letter variable?

glouppe replied:
I am keeping that one. This is how it is defined in the parent `_BaseKFold` class. I am not sure it is a good idea to change the API, even if indeed `n_samples` would be clearer.

jnothman replied:
Surely both these assignments are unnecessary as they're done by parent classes?

arjoly commented on sklearn/cross_validation.py:
```diff
@@ -339,6 +342,121 @@ def __len__(self):
+        # Weight labels by their number of occurences
+        samples_per_label = np.bincount(labels)
+
+        # Distribute the most frequent labels first
```
You mean in decending order?

arjoly commented on sklearn/cross_validation.py:
```diff
@@ -339,6 +342,121 @@ def __len__(self):
+        samples_per_label = np.bincount(labels)
+
+        # Distribute the most frequent labels first
+        ind = np.argsort(samples_per_label)[::-1]
```
ind -> index

arjoly commented on sklearn/cross_validation.py:
```diff
@@ -339,6 +342,121 @@ def __len__(self):
+
+        # While there are weights, distribute them
+        # Specifically, add the biggest weight to the lightest fold
+        for label_index, w in enumerate(samples_per_label):
```
w -> weight?

arjoly replied:
(to avoid one letter variable)

arjoly commented on sklearn/cross_validation.py:
```diff
@@ -339,6 +342,121 @@ def __len__(self):
+        samples_per_label = samples_per_label[ind]
+
+        # Total weight of each fold
+        samples_per_fold = np.zeros(n_folds, dtype=np.uint64)
```
samples_per_fold -> n_samples_per_fold?

arjoly replied:
I am not sure this should be dtyped.

arjoly commented on sklearn/cross_validation.py:
```diff
@@ -339,6 +342,121 @@ def __len__(self):
+        # While there are weights, distribute them
+        # Specifically, add the biggest weight to the lightest fold
+        for label_index, w in enumerate(samples_per_label):
+            ind_min = np.argmin(samples_per_fold)
```
I would try to find a more descriptive name.

arjoly commented on sklearn/cross_validation.py:
```diff
@@ -339,6 +342,121 @@ def __len__(self):
+            samples_per_fold[ind_min] += w
+            label_to_fold[ind[label_index]] = ind_min
+
+        self.idxs = label_to_fold[labels]
```
idxs -> indexs or something more coherent with the rest of the module?

glouppe replied:
Same here, this comes from the `_BaseKFold` API.

arjoly commented on sklearn/cross_validation.py:
```diff
@@ -339,6 +342,121 @@ def __len__(self):
+
+    Examples
+    --------
+    >>> from sklearn import cross_validation
```
I would directly import LabelKFold.

arjoly commented:
I made some nitpicks, but overall the code looks good.

The naming questions shouldn't be neglected.

glouppe commented:
Indeed the code needs some polishing (I kept most of #4444), I'll address your comments tomorrow.

glouppe committed:
COSMIT: variable names, documentation, etc

glouppe commented:
@arjoly I think a addressed all your comments.

arjoly commented:
thanks!

jnothman commented:
LGTM... Are we certain we need this and the new `LabelShuffleSplit`, though?

JeanKossaifi commented:
Yes, they have different uses: LabelShuffleSplit doesn't maintain balance between folds whereas this class creates K disjoint approximately equilibrated folds usable, for instance for person specific K-fold cross validation.

jnothman commented:
Well apart from that nitpick about attribute initialisation above, I think we can merge this...

glouppe committed:
ENH: remove unnecessary assignments

glouppe commented:
@jnothman Indeed, thanks for noticing! This has been fixed.

I added a what's new entry. Waiting for CI to give the green light to merge.
